### PR TITLE
feat(lsp): overwrite omnifunc/tagfunc set by ftplugin

### LINKED ISSUE #21001: Use vim.lsp.omnifunc instead of the built-in one
### Feature already in Vim?

_No response_

### Feature description

PR #19003 introduced a feature which automatically sets `omnifunc` to the one provided by the language server if it is not set.

The goal was to reduce the boilerplate code in user configuration:
> This is boilerplate that we may be able to avoid if we can come up with a choice that is uncontroversial enough. I think it would be acceptable if it removes boilerplate for ~90% of the use-cases, while maybe adding a bit of extra config for the remaining 10%.

The problem is that for many languages the ftplugin already sets the completion function to a built-in one which is (most of the time?) worse than the LSP-based one. For example, C++ has `ccomplete#Complete` which is for C and does not work at all with C++. And in this case the LSP on_attach callback will not set `omnifunc` because it is already set.

List of impacted languages:
```
$ rg -l 'setlocal (omnifunc|ofu)' runtime/ftplugin
runtime/ftplugin/html.vim
runtime/ftplugin/zimbu.vim
runtime/ftplugin/cucumber.vim
runtime/ftplugin/xml.vim
runtime/ftplugin/mp.vim
runtime/ftplugin/mf.vim
runtime/ftplugin/php.vim
runtime/ftplugin/sql.vim
runtime/ftplugin/context.vim
runtime/ftplugin/sass.vim
runtime/ftplugin/haskell.vim
runtime/ftplugin/c.vim
runtime/ftplugin/ada.vim
runtime/ftplugin/python.vim
runtime/ftplugin/debchangelog.vim
runtime/ftplugin/ruby.vim
runtime/ftplugin/less.vim
runtime/ftplugin/javascript.vim
runtime/ftplugin/css.vim
```

Would it be possible/acceptable to change the logic introduced in #19003 and take into account the built-in `omnifunc`s and overwrite them as well?

### PR DESCRIPTION
**Pull request by mliszcz**
Some built-in ftplugins set omnifunc/tagfunc/formatexpr which prevents using the LSP-based ones as default. This change allows for overwriting omnifunc/tagfunc/formatexpr options also set by ftplugin with functions provided by the LSP client.

Fixes #21001.

For example, for C++ we have:
```
  omnifunc=ccomplete#Complete                                                                                                                                                        
        Last set from /usr/share/nvim/runtime/ftplugin/c.vim line 30
```
And the changes done in #19003 will never apply to this language.

### CODE REVIEW

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1093,19 +1093,29 @@ function lsp.start_client(config)
...
   end
 
+  ---@private
+  -- Determines whether the given option can be set by `set_defaults`.
+  local function is_empty_or_default(bufnr, option)
+    if vim.bo[bufnr][option] == '' then
+      return true
+    end
+    local last_set_from = vim.fn.gettext('\n\tLast set from ')
+    return not not vim.fn.execute('verbose set ' .. option .. '?'):match(last_set_from .. '.*ftplugin')
```
Comment (mliszcz) [inline runtime/lua/vim/lsp.lua:1103]: I'm wondering if we should be more strict here, perhaps we could match against '.*/runtime/ftplugin' or even $VIMRUNTIME because ftplugin may be part of user's configuration.

Comment (glepnir) [inline runtime/lua/vim/lsp.lua:1103]: I wonder  does there can not simply use `nvim_get_option_value` ?

Comment (mliszcz) [reply runtime/lua/vim/lsp.lua:1103]: Nope, this does not provide information from where the option was set. I think that going through vimscript is currently the only way.

Comment (clason) [reply runtime/lua/vim/lsp.lua:1103]: it does, if you use `-V1`

Comment (glepnir) [reply runtime/lua/vim/lsp.lua:1103]: did you try use  lua file in`ftplugin`?  i guess this way might not works fine when it's lua file. you will get message like `last set from lua`

Comment (mliszcz) [reply runtime/lua/vim/lsp.lua:1103]: @clason
Where should `-V1` go? This is not mentioned in the docs (https://neovim.io/doc/user/api.html).
The second parameter (opts) is either scope, buf or win.
```
:lua print(vim.api.nvim_get_option_value('omnifunc', {})) 
```

@glepnir
Yes it if is set from lua file it does work as you say. However, the goal of this PR is to overwrite only values set by builtin ftplugins shipped with neovim. All of them are implemented in vimscript right now. I think those are inherited from vim and will stay as vimscript files, right? You can find the list of impacted languages in the corresponding issue.

Comment (clason) [reply runtime/lua/vim/lsp.lua:1103]: No, I mean that if you run Nvim with `-V1`, you _do_ get information where the option was set from Lua. This is not sufficient justification to use vimscript here.

**Comment by clason** [comment]
@mfussenegger

File: runtime/lua/vim/lsp.lua (hunk 1 above)
Comment (justinmk) [reply runtime/lua/vim/lsp.lua:1103]: checking `$VIMRUNTIME` seems like a good idea.

Also need a test exercising this. Hopefully just add some checks in here : https://github.com/neovim/neovim/blob/2e450efb9525308708dd9e26d6ea1d029ac781b8/test/functional/plugin/lsp_spec.lua#L278 

or, alas, maybe need to copy-paste that test and create a `it('sets defaults')` test.

Comment (zeertzjq) [reply runtime/lua/vim/lsp.lua:1103]: A Vim patch may be relevant, combined with `nvim_get_option_info`: https://github.com/vim/vim/commit/f768c3d19c518822d89dec4cc3947ddeea249316

**Comment by mliszcz** [comment]
Thanks all for your comments so far. I started adding functional tests and encountered a blocking problem: `verbose set omnifunc?` gets the option from the currently selected buffer. No buffer is selected in the tests. I'm not sure if this can also happen during normal operation. In any case I don't think that switching buffers in the on_attach callback is a good idea.

The mentioned `nvim_get_option_info` will also not work. It queries the global `options` table and while 'omnifunc' and others are listed there, their `last_set` field is never updated (https://github.com/neovim/neovim/blob/master/src/nvim/option.c#L1935).

I think that it would be best to add a new API function `nvim_**buf**_get_option_info` that would also accept a bufnr. Then probably we will need the vim patch mentioned by zeertzjq to resolve the sid to a filename. What do you think?

One more argument for adding a new API function is that `verbose set?` output is problematic to parse, e.g.:

```
echo 'set omnifunc=test' > 'cpp.vim line 100'
```
```
:source cpp.vim\ line\ 100
:verbose set omnifunc? 
  omnifunc=test                                                                                                                                                                      
        Last set from ~/var/neovim/cpp.vim line 100 line 1
```

**Comment by justinmk** [comment]
> add a new API function `nvim_**buf**_get_option_info` that would also accept a bufnr.

The `nvim_buf`/`nvim_win` aren't a good match for option scopes because of the complexity of option scopes. Would need a `opts` param, interface similiar to `nvim_get_option_value`.



> `verbose set?` output is problematic to parse, e.g.:

The check is "best effort" anyway. The stakes are low, we're just making a housekeeping attempt to decide whether to set an option. We can document this for now.

**Comment by mliszcz** [comment]
> > add a new API function `nvim_**buf**_get_option_info` that would also accept a bufnr.
> 
> `nvim_buf_*`/`nvim_win_*` aren't a good match for option scopes because of the complexity of option scopes. Would need a `opts` param, interface similiar to `nvim_get_option_value`.

If you agree I can add that, even in this PR in a separate commit. Just please tell what would be the preferred approach:
1. add a mandatory `opts` parameter to the C implementation of `nvim_get_option_info(name, opts)`. This would break the API,  Lua users would need to pass an empty `{}`,
2. create a new API function that would accept both name and opts, perhaps deprecate the original one,
3. add an optional `opts` parameter to `nvim_get_option_info(name, opts)`:
    * rename the C function to e.g. `_nvim_get_option_info_impl` and add mandatory `opts` parameter
    * in  `_meta.lua` define `nvim_get_option_info` that would check if `opts` is present and either pass it to the C function or use a `{}` as a default value.
4. something else?

**Comment by clason** [comment]
> add a mandatory opts parameter to the C implementation of nvim_get_option_info(name, opts). This would break the API, Lua users would need to pass an empty {},

That's a no-go, according to `:h api-contract`. Would need to make it a new function with a (slightly) different name and deprecate the old. (This Is The Way.)

**Comment by justinmk** [comment]
> If you agree I can add that, even in this PR in a separate commit

Let's finish this PR first, then definitely welcome to explore that in a new PR. Also, what @clason said :)

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -1112,19 +1112,45 @@ function lsp.start_client(config)
...
+  local function is_empty_or_default(bufnr, option)
+    if vim.bo[bufnr][option] == '' then
+      return true
+    end
+
+    local old_bufnr = vim.fn.bufnr('')
+    local last_set_from = vim.fn.gettext('\n\tLast set from ')
+    local line = vim.fn.gettext(' line ')
+
+    vim.fn.execute('buffer ' .. bufnr)
```
Comment (craigmac) [inline runtime/lua/vim/lsp.lua:1126]: Are we able to use `vim.cmd.buffer(bufnr)` instead, here and below?

Comment (mliszcz) [reply runtime/lua/vim/lsp.lua]: Yes, done. Thanks!

File: runtime/lua/vim/lsp.lua (hunk 1 above)
Comment (mliszcz) [reply runtime/lua/vim/lsp.lua]: I added tests and changed the code to check for vimruntime.

**Comment by mliszcz** [comment]
Hi all. I just added some tests and addressed all open issues. I believe the PR is ready for the next review round. Thanks in advance!

**Comment by mliszcz** [comment]
Kindly reminder: all comments were addressed, please have a look.

I'm pinging everyone involved so far, in case you have some more comments.
@clason
@craigmac
@zeertzjq
@glepnir
@justinmk

File: test/functional/plugin/lsp_spec.lua [hunk 3]
```diff
@@ -30,6 +30,11 @@ local fake_lsp_code = lsp_helpers.fake_lsp_code
 local fake_lsp_logfile = lsp_helpers.fake_lsp_logfile
 local test_rpc_server = lsp_helpers.test_rpc_server
 
+local function get_buf_option(name, bufnr)
+    bufnr = bufnr or "BUFFER"
+    return exec_lua(string.format("return vim.api.nvim_buf_get_option(%s, '%s')", bufnr, name))
```
Comment (justinmk) [inline test/functional/plugin/lsp_spec.lua:35]: This indirection makes the tests hard to follow. But I don't see an obvious way to avoid this.

**Review by justinmk — APPROVED** [review]
LGTM. It's a lot of code unfortunately, but the tests look thorough, so thank you!

We'll need to keep in mind `is_empty_or_default` and see if we can follow up with a less ad-hoc approach (`nvim_get_option_info` as discussed) or possibly extract it to a util somewhere. Ideally vim would provide better support for this (maybe a recent patch?), similar to how `mapset()`/`maparg()` were added to better support user-defined mappings.

**Comment by Ursescu** [comment]
Hi, it seems that this is throwing some errors in combinations with some third party plugins (Nvimtree && auto-sessions, but I don't this it's that important). It may happen that old_buffer is deleted when entering to `bufnr` in this `lsp.lua` section of `is_empty_or_default`:
```
    vim.cmd.buffer(bufnr)
    local scriptname = vim.fn
      .execute('verbose set ' .. option .. '?')
      :match(last_set_from .. '(.*)' .. line .. '%d+')
    vim.cmd.buffer(old_bufnr) -- old_bufnr doesn't exist anymore
 ```
 
 Some logs:
 ```
 Buffers before changing buffer: 
{ 2, 3, 4, 5, 6 }
2                                                                                                                                                                                                                   
3 init.lua                                                                                                                                                                                                           
4 /home/john/.config/nvim/lua/keymaps.lua                                                                                                                                                                            
5 lua/config/auto-session.lua                                                                                                                                                                                        
6 NvimTree_1                                                                                                                                                                                                         
    
old_buffer is now -> 6
Buffers after changing buffer:
{ 2, 3, 4, 5 }  -> 6 deleted                                                                                                                                                                                                                                                                                                                                                                                                    
2                                                                                                                                                                                                                   
3 init.lua                                                                                                                                                                                                           
4 /home/john/.config/nvim/lua/keymaps.lua                                                                                                                                                                            
5 lua/config/auto-session.lua                                                                                                                                                                                        
                                                                                                                                                                                                     
Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/lsp.lua:1121: Vim:E86: Buffer 6 does not exist                                                                                           
stack traceback:                                                                                                                                                                                                    
        [C]: in function 'buffer'                                                                                                                                                                                   
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1121: in function 'is_empty_or_default'                                                                                                                             
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1133: in function 'set_defaults'                                                                                                                                    
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1546: in function '_on_attach'                                                                                                                                      
        /usr/share/nvim/runtime/lua/vim/lsp.lua:1375: in function ''                                                                                                                                                
        vim/_editor.lua: in function <vim/_editor.lua:0>  
 ```

**Comment by mliszcz** [comment]
Hi, I was not aware that some other code could be executed between storing and switching back to the buffer. I think that the only option would be to just ignore this error (`silent!`). We cannot check if the buffer still exists because it could be deleted after the check as well, right? Also this is still not good because we will end up switching to (staying in) some arbitrary buffer instead of the current one.

In any case I am working on improving this by extending nvim_get_option_info so that we do not need to switch buffers at all. A PR should be ready by the end of this week.

It would be best if one of the maintainers could comment on what we should do with this.

File: runtime/lua/vim/lsp.lua [hunk 4]
```diff
@@ -1112,19 +1112,43 @@ function lsp.start_client(config)
...
+
+    local old_bufnr = vim.fn.bufnr('')
+    local last_set_from = vim.fn.gettext('\n\tLast set from ')
+    local line = vim.fn.gettext(' line ')
+
+    vim.cmd.buffer(bufnr)
+    local scriptname = vim.fn
+      .execute('verbose set ' .. option .. '?')
+      :match(last_set_from .. '(.*)' .. line .. '%d+')
+    vim.cmd.buffer(old_bufnr)
```
[the remark this answers is not in the document]
Comment (mliszcz) [reply runtime/lua/vim/lsp.lua:1130]: Should I send a PR with this change?

[the remark this answers is not in the document]
Comment (mliszcz) [reply runtime/lua/vim/lsp.lua:1130]: Well I guess this depends on what is causing the buffer to be deleted. @Ursescu would you be able to build and test if I open a PR? Or alternatively could you please share a minimal vimrc and steps needed to reproduce the problem?

[the remark this answers is not in the document]
Comment (Ursescu) [reply runtime/lua/vim/lsp.lua:1130]: I copy-pasted the `noautocmd` variant, but it doesn't work... I am not sure why, because this made sense. I will investigate further and get back to you.

**Comment by justinmk** [comment]
coincidentally, https://github.com/neovim/neovim/pull/22634 implements `vim.filetype.get_option()` which should greatly simplify `is_empty_or_default` here

File: runtime/lua/vim/lsp.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ursescu) [reply runtime/lua/vim/lsp.lua:1130]: After doing some tests I concluded that the command itself will cause the NvimTree buffer to be unloaded. I am not sure how `:buffer` really works under the hood and what interactions are between unlisted buffers and normal buffers. Even if I wrap the whole thing with `eventignore` to stop the events (I actually added prints on each autocmd registered and none is triggered), still the buffer is unloaded. I tried the command `:buffer 4` (some random existing buffer) while current buffer == NvimTree buffer and I see the same behavior (buffer unloaded).
```
:buffers!                                                                                                                                                                                                           
  2u h   "[Scratch]"                    line 0                                                                                                                                                                      
  3  a   "init.lua"                     line 175                                                                                                                                                                    
  4      "stuff.txt"                    line 1                                                                                                                                                                      
  5      "lua/autocmds.lua"             line 27                                                                                                                                                                     
  6 #a   "~/.config/nvim/lua/plugins.lua" line 101                                                                                                                                                                  
  8u%a-  "NvimTree_1"                   line 4  
  ```
  So in the end I think the issue is more or less dependent on the setup/plugins/configuration and not necessarily due to this PR. Still it would be awesome to not rely on buffer switching here.

[the remark this answers is not in the document]
Comment (craigmac) [reply runtime/lua/vim/lsp.lua:1130]: try `:verbose 15` to see what autocmds are being fired as they happen, might help with narrowing it down
